lib/sms/sms.c

```
@@ -0,0 +1,322 @@
+			ret);
+		return ret;
+	}
+
+	ret = at_parser_max_params_from_str(resp, NULL, &resp_list,
+					    AT_CNMI_PARAMS_COUNT);
+	if (ret) {
+		LOG_INF("%s", log_strdup(resp));
+		LOG_ERR("Invalid AT response, err: %d", ret);
+		return ret;
+	}
+
+	int cnmi = 0, value;
+
+	for (int i = 1; i < AT_CNMI_PARAMS_COUNT - 1; i++) {
+		ret = at_params_int_get(&resp_list, i, &value);
+		if (ret) {
+			LOG_ERR("Invalid AT response parameters, err: %d", ret);
+			return ret;
+		}
+		cnmi += value;
+	}
+
+	/* No client are registered when the first 4 parameters are 0. */
+	if (cnmi != 0) {
+		LOG_ERR("Only one SMS client can be registred");
+		return -EBUSY;
+	}
+
+	/* Register for AT commands notifications before creating the client. */
+	ret = at_notif_register_handler(NULL, sms_at_handler);
+	if (ret) {
+		LOG_ERR("Cannot register AT notification handler, err: %d",
+			ret);
+		return ret;
+	}
+
+	/* Register this module as an SMS client. */
+	ret = at_cmd_write(AT_SMS_SUBSCRIBER_REGISTER, NULL, 0, NULL);
+	if (ret) {
+		(void)at_notif_deregister_handler(NULL, sms_at_handler);
+		LOG_ERR("Unable to register a new SMS client, err: %d", ret);
+		return ret;
+	}
+
+	/* Clear all observers. */
+	for (size_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
+
+		s->ctx = NULL;
+		s->listener = NULL;
+	}
+
+	sms_client_registered = true;
+	LOG_INF("SMS client successfully registered");
+	return 0;
+}
+
+sms_handle_t sms_register_listener(sms_callback_t listener, void *context)
+{
+	if (listener == NULL) {
+		return -1; /* Invalid parameter. */
+	}
+
+	/* Search for a free slot to register a new listener. */
+	for (sms_handle_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
+
+		if (s->ctx == NULL && s->listener == NULL) {
+			s->ctx = context;
+			s->listener = listener;
+			return i;
+		}
+	}
+
+	/* Too many subsribers. */
+	return -1;
```

Shall we return the actual error code here? The function itself is pretty straightforward, still it would be good to differentiate between `-EINVAL`/`-ENOMEM` here.

---

Good point. Fixed.